src/scf/potential_mixing.jl

```diff
@@ -0,0 +1,345 @@
+using Statistics
+
+# Quick and dirty Anderson implementation ... lacks important things like
+# control of the condition number of the anderson matrix.
+# Not particularly optimised. Also should be moved to NLSolve ...
+function AndersonAcceleration(;m=Inf)
+    # Accelerates the iterative solution of f(V) = 0, where for our DFT case:
+    #    f(V) = Vext + Vhxc(ρ(V)) - V
+    # Further define
+    #    preconditioned update    Pf(V) = P⁻¹ f(V)
+    #    fixed-point map          g(V)  = V + α Pf(V)
+    # where the α may vary between steps.
+    #
+    # Finds the linear combination Vₙ₊₁ = g(Vₙ) + ∑ᵢ βᵢ (g(Vᵢ) - g(Vₙ))
+    # such that |Pf(Vₙ) + ∑ᵢ βᵢ (Pf(Vᵢ) - Pf(Vₙ))|² is minimal
+    #
+    Vs   = []  # The V     for each iteration
+    PfVs = []  # The Pf(V) for each iteration
+
+    function extrapolate(Vₙ, αₙ, PfVₙ)
+        m == 0 && return Vₙ .+ αₙ .* PfVₙ
+
+        # Gets the current Vₙ, Pf(Vₙ) and damping αₙ
+        #
+        Vₙ₊₁ = vec(Vₙ) .+ αₙ .* vec(PfVₙ)
+        if !isempty(Vs)
+            M = hcat(PfVs...) .- vec(PfVₙ)  # Mᵢⱼ = (PfVⱼ)ᵢ - (PfVₙ)ᵢ
+            # We need to solve 0 = M' PfVₙ + M'M βs <=> βs = - (M'M)⁻¹ M' PfVₙ
+            βs = -M \ vec(PfVₙ)
+            for (iβ, β) in enumerate(βs)
+                Vₙ₊₁ .+= β .* (Vs[iβ] .- vec(Vₙ) .+ αₙ .* (PfVs[iβ] .- vec(PfVₙ)))
+            end
+        end
+
+        push!(Vs,   vec(Vₙ))
+        push!(PfVs, vec(PfVₙ))
+        if length(Vs) > m
+            Vs   = Vs[2:end]
+            PfVs = PfVs[2:end]
+        end
+        @assert length(Vs) <= m
+
+        reshape(Vₙ₊₁, size(Vₙ))
+    end
+end
+
+
+"""
+Accept a step if the energy is at most increasing by `max_energy` and the residual
+is at most `max_relative_residual` times the residual in the previous step.
+"""
+function ScfAcceptImprovingStep(;max_energy_change=1e-12, max_relative_residual=0.9)
```

Comment:
This pattern of `function CallbackFunction(; params) info -> do_stuff(info, params)`, which we also use in scf_callbacks, is kind of solidifying, we should think about it a bit. Hiding params (or worse, state) in closures is not so great. Alternatives are
`struct CallbackFunction params end; do_stuff(::CallbackFunction, info)` or `struct CallbackFunction params end; (::CallbackFunction)(info)`. I tend to prefer the third, thoughts?

Reply:
I agree. Thinking about the Anderson I also lean towards 2 or 3 as better Options. I think which is better depends a bit on the use case. What I don't like about the third option in general is that it suggests the struct is only meant to do one thing. For callbacks that is, however, the standard use case, so perhaps most can be converted to that pattern without much trouble.